src/rest/RequestHandler.js

```diff
@@ -0,0 +1,174 @@
+const DiscordAPIError = require('./DiscordAPIError');
+const HTTPError = require('./HTTPError');
+const Util = require('../util/Util');
+const { Events: { RATE_LIMIT }, browser } = require('../util/Constants');
+
+function parseResponse(res) {
+  if (res.headers.get('content-type').startsWith('application/json')) return res.json();
+  // eslint-disable-next-line no-undef
+  if (browser) return res.blob();
+  return res.buffer();
+}
+
+class RequestHandler {
+  constructor(manager) {
+    this.manager = manager;
+    this.busy = false;
+    this.queue = [];
+    this.reset = -1;
+    this.remaining = -1;
+    this.limit = -1;
+    this.retryAfter = -1;
+  }
+
+  push(request) {
+    this.queue.push(request);
+    this.run();
+  }
+
+  run() {
+    if (this.queue.length === 0) return;
+    this.execute(this.queue.shift());
+  }
+
+  get limited() {
+    return (this.manager.globallyRateLimited || this.remaining <= 0) && Date.now() < this.reset;
+  }
+
+  get _inactive() {
+    return this.queue.length === 0 && !this.limited && this.busy !== true;
+  }
+
+  _getAPIOffset(serverDate) {
+    return new Date(serverDate).getTime() - Date.now();
+  }
+
+  _calculateReset(reset, serverDate) {
+    return new Date(Number(reset) * 1000).getTime() - this._getAPIOffset(serverDate);
+  }
+
+  execute(item) {
+    // Insert item back to the beginning if currently busy
+    if (this.busy) {
+      this.queue.unshift(item);
+      return null;
+    }
+
+    this.busy = true;
+
+    // Perform the request
+    return item.request.make()
+      .then(async res => {
+        if (res && res.headers) {
+          if (res.headers.get('x-ratelimit-global')) this.manager.globallyRateLimited = true;
+
+          const serverDate = res.headers.get('date');
+          const limit = res.headers.get('x-ratelimit-limit');
+          const remaining = res.headers.get('x-ratelimit-remaining');
+          const reset = res.headers.get('x-ratelimit-reset');
+          const retryAfter = res.headers.get('retry-after');
+
+          this.limit = limit ? Number(limit) : Infinity;
+          this.remaining = remaining ? Number(remaining) : 1;
+          this.reset = reset ? this._calculateReset(reset, serverDate) + 100 : Date.now();
```

Comment:
why are we offsetting reset times by 100ms

Reply:
I figured, as before, we should have some leeway in case we are ever barely off in calculations. Is it better removed entirely, or reduced?

Reply:
Removed entirely. We already have a rainy day fund that no other discord library implements (let alone has on by default), which is d.js reactions are slower than any other library, even those who don't hardcode. ClientOptions#restTimeOffset... Speaking of, it looks like you have inadvertently removed that functionality.

A rainy day fund is really not necessary, but if it's there, it should continue to be configurable.